lib/dao.js

```diff
@@ -587,6 +626,177 @@ DataAccessObject.updateOrCreate = DataAccessObject.upsert = function upsert(data
+ *
+ * NOTE: No setters, validations, or hooks are applied when using upsert.
+ * @param {Object} data The model instance data
+ * @param {Object} [options] Options for replaceOrCreate
+ * @param {Function} cb The callback function (optional).
+ */
+
+DataAccessObject.replaceOrCreate  = function replaceOrCreate(data, options, cb) {
+  var connectionPromise = stillConnecting(this.getDataSource(), this, arguments);
+  if (connectionPromise) {
+    return connectionPromise;
+  }
+
+  if (options === undefined && cb === undefined) {
+    if (typeof data === 'function') {
+      cb = data;
+      data = {};
+    }
+  } else if (cb === undefined) {
+    if (typeof options === 'function') {
+      cb = options;
+      options = {};
+    }
+  }
+
+  cb = cb || utils.createPromiseCallback();
+  data = data || {};
+  options = options || {};
+
+  assert(typeof data === 'object', 'The data argument must be an object');
+  assert(typeof options === 'object', 'The options argument must be an object');
+  assert(typeof cb === 'function', 'The cb argument must be a function');
+
+  var hookState = {};
+
+  var self = this;
+  var Model = this;
+  var connector = Model.getConnector();
+
+  var id = getIdValue(this, data);
+  if (id === undefined || id === null) {
+    return this.create(data, options, cb);
+  }
+  
+  var inst;
+  if (data instanceof Model) {
+    inst = data;
+  } else {
+    inst = new Model(data);
+  }
+  
+  var context = {
+    Model: Model,
+    query: byIdQuery(Model, id),
+    hookState: hookState,
+    options: options
+  };
+  Model.notifyObserversOf('access', context, doReplaceOrCreate);
+
+  function doReplaceOrCreate(err, ctx) {
+    if (err) return cb(err);
+
+    var isOriginalQuery = isWhereByGivenId(Model, ctx.query.where, id);
+    var where = ctx.query.where;
+    if (connector.replaceOrCreate && isOriginalQuery) {
+      var context = {
+        Model: Model,
+        instance: inst,
+        hookState: hookState,
+        options: options
+      };
+      Model.notifyObserversOf('before save', context, function(err, ctx) {
+        if (err) return cb(err);
+
+        var update = inst.toObject(false);
+        Model.applyProperties(update, inst);
+        Model = Model.lookupModel(update);
+
+        var connector = self.getConnector();
+
+        inst.isValid(function(valid) {
+          if (!valid) {
+              return cb(new ValidationError(inst), inst);
+          }
+          callConnector();
+        }, update);
+
+        function callConnector() {
+          update = removeUndefined(update);
+          context = {
+            Model: Model,
+            where: where,
+            data: update,
+            currentInstance: inst,
+            hookState: ctx.hookState,
+            options: options
+          };
+          Model.notifyObserversOf('persist', context, function(err) {
+            if (err) return done(err);
+            connector.replaceOrCreate(Model.modelName, update, options, done);
```

Comment:
I think you should use `context.data` instead of `update` to make the following hook work:

```js
Model.observe('persist', function(ctx, next) {
  ctx.data = { newdata: true };
});
```